### PR TITLE
fix: Removed unnecessary rendering of microphone permission warning info.

### LINKED ISSUE #7403: [BUG]: Unnecessary rendering Mic Permission warning even if mic permission is enabled
**Describe the bug**

In the audio recording section of the Upload Sample report, a warning message appears stating, "Please allow browser permission before you start speaking," even when permission is already enabled or when the recording is taking place, which is inappropriate.

**To Reproduce**

1. Go to Sample Test.
2. Scroll down to find the Upload Sample Report header title.
3. Notice the unnecessary presence of the warning message.

**Expected behavior**

The warning message should only appear when the recording fails due to the lack of microphone permission.

**Screenshots / Videos**

https://github.com/coronasafe/care_fe/assets/78961432/1d5c740a-0338-4339-b98d-996b8524eb3d

**Desktop:**

- OS: Ubuntu 22.04.3 LTS
- Browser: Microsoft Edge
- Version: 120.0.2210.133 (Official build) (64-bit)

### PR DESCRIPTION
**Pull request by Spiral-Memory**
## Proposed Changes

- Fixes #7403 
- It assumes that mic permission is enabled at first.
- Added a state to indicate if Mic Permission is enabled.
- Based on this state, the warning is rendered.
- If permission is not given and the `requestRecorder()` function throws an error, then we set the state accordingly.

@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

## Merge Checklist

- [x] Add specs that demonstrate the bug / test a new feature.
- [ ] Update [product documentation](https://docs.coronasafe.network/coronasafe-care-documentation/architecture/architecture-and-layering-of-care).
- [x] Ensure that UI text is kept in I18n files.
- [x] Prepare a screenshot or demo video for the changelog entry, and attach it to the issue.
- [ ] Request Peer Reviews
- [ ] Completion of QA

https://github.com/coronasafe/care_fe/assets/78961432/eee43fee-098f-4cdb-b823-8cfab03bda3c

### CODE REVIEW

**Comment by Spiral-Memory** [comment]
@rithviknishad , please take a look at this PR. In the CodeClimate test, it is showing that a type definition should be included, but the file I changed is a JavaScript file. If I add a type definition, it will likely result in an error. Kindly review this PR, investigate this issue, and let me know. Thanks.

**Review by Ashesh3 — APPROVED** [review]

**Comment by Spiral-Memory** [comment]
Hey @rithviknishad Could you please look into this once ?

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
image

**Comment by Spiral-Memory** [comment]
> image
> 

Could you please explain a little, like what's the problem ? 

When permission is not there.. that warming will be shown otherwise it's not. Please refer to the video on how i intended it to be. 

If you require any change, please let me know !

**Comment by rithviknishad** [comment]
Hey @Spiral-Memory 

Let us just confirm that doing this is okay as IIRC this was kept intentionally, don't remember the reason.

**Comment by Spiral-Memory** [comment]
> Hey @Spiral-Memory 
> 
> Let us just confirm that doing this is okay as IIRC this was kept intentionally, don't remember the reason.

Sure @rithviknishad !

**Comment by gigincg** [comment]
@Spiral-Memory @nihal467 @rithviknishad 

Following is the expected Behaviour:

Warning as span :

- Shown when Permission is not granted
- Hidden when Permission is already granted

Warning as notification:

- Shown when Permission is not granted & User clicks on `Record` button

**Comment by Spiral-Memory** [comment]
> @Spiral-Memory @nihal467 @rithviknishad
> 
> Following is the expected Behaviour:
> 
> Warning as span :
> 
> * Shown when Permission is not granted
> * Hidden when Permission is already granted
> 
> Warning as notification:
> 
> * Shown when Permission is not granted & User clicks on `Record` button

Yep! This is exactly what's happening after this PR? Is there any unexpected behavior? Please let me know. I will fix it

**Comment by Spiral-Memory** [comment]
Okay, Got it, currently the span warning is not getting rendered if mic permission is not there and is rendering after clicking the button. 

Let me make another commit to fix it.

**Comment by Spiral-Memory** [comment]
Have a look now @rithviknishad @gigincg @Ashesh3 @nihal467 
Exactly as per the behavior mentioned.

https://github.com/coronasafe/care_fe/assets/78961432/c48dcfb4-ee01-46ef-9e6d-d9ba3ac17e0f

**Comment by nihal467** [comment]
@Spiral-Memory 
image

- [ ] When you go to a patient consultation page and click on the "Files" options in the navigation bar, you will see Chrome asking the user to allow the use of the mic. This shouldn't be the case. It is expected to ask the user for mic permission only when they click on the record button in the "Files" section.

**Comment by Spiral-Memory** [comment]
@nihal467 , I tried using the query now instead of getUserMedia to avoid the permission popup. Please verify if it works fine now.

The following are the behaviors now:

1. If mic permission is not there or blocked, the span will be visible to allow mic permission.
2. In case permission is not there and only when the user clicks on 'Start Record', the popup will appear asking for mic permission.
3. In case permission is blocked, the popup won't appear, and the span warning plus a toast message will be displayed.
4. If mic permission is allowed, there won't be such a span warning."

File: src/Components/Patient/FileUpload.tsx [hunk 1]
```diff
@@ -232,6 +232,30 @@ export const FileUpload = (props: FileUploadProps) => {
...
+    const checkMicPermission = async () => {
+      try {
+        const permissions = await navigator.permissions.query({
+          name: "microphone" as PermissionName,
+        });
+        if (permissions.state === "granted") {
+          setIsMicPermission(true);
+        } else {
+          setIsMicPermission(false);
+        }
```
Comment (rithviknishad) [inline src/Components/Patient/FileUpload.tsx]:
```suggestion
        setIsMicPermission(permissions.state === "granted");
```

File: src/Components/Patient/FileUpload.tsx [hunk 2]
```diff
@@ -564,6 +588,12 @@ export const FileUpload = (props: FileUploadProps) => {
     setbtnloader(false);
   };
 
+  const handleSetMicPermission: (isPermitted: boolean) => void = (
+    isPermitted
+  ) => {
+    setIsMicPermission(isPermitted);
+  };
+
```
Comment (rithviknishad) [inline src/Components/Patient/FileUpload.tsx]:
```suggestion
```

File: src/Components/Patient/FileUpload.tsx [hunk 3]
```diff
@@ -1510,8 +1540,9 @@ export const FileUpload = (props: FileUploadProps) => {
                         confirmAudioBlobExists={confirmAudioBlobExists}
                         reset={resetRecording}
                         setResetRecording={setResetRecording}
-                      />
-                      {!audioBlobExists && (
+                        handleSetMicPermission={handleSetMicPermission}
```
Comment (rithviknishad) [inline src/Components/Patient/FileUpload.tsx]:
```suggestion
                        handleSetMicPermission={setIsMicPermission}
```

**Review by rithviknishad — COMMENTED** [review]
LGTM, just some minor suggestions

**Comment by rithviknishad** [comment]
Also fix the lint errors

**Review by Ashesh3 — APPROVED** [review]

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
lgtm

**Review by khavinshankar — APPROVED** [review]

**Pull request merged**